feat: locale sync check GH action (from #1297, fixes #1444)

rishi-jat opened the pull request:
## 🌐 Locale Sync Check GitHub Action

   **Fixes:** #1444
   **Related:** #1297 (closed PR that this completes)

   ### Changes Made
   - ✅ Added locale synchronization check script (`.github/scripts/check-locale-sync.js`)
   - ✅ Added GitHub Actions workflow (`.github/workflows/locale-sync.yml`)
   - ✅ Added test value to English locale file (`frontend/src/locales/strings.en.json`)
   - ✅ Added `check-locale-sync` script to `frontend/package.json`

   ### Features
   - 🔍 Automatic detection of missing/extra translation keys
   - 📝 GitHub issue creation and management for translation discrepancies
   - 🔄 PR integration with automatic commenting
   - 🎯 Self-healing: closes issues when translations are fixed
   - 📊 Comprehensive reporting and status tracking

   This completes the localization check GitHub action work that was started in PR #1297.

clubanderson commented:
@manzil-infinity180 please advise on trivy fail

/ok-to-test

manzil-infinity180 commented:
Rebase it with the latest change

rishi-jat commented:
Rebase with the latest dev branch is done as requested.✅
@manzil-infinity180  please have a look when you get a chance.
If all looks good, kindly add /lgtm and /approve.

manzil-infinity180 commented:
```
 - name: Run locale sync check
        id: locale-check
        run: |
          cd frontend
          npm run check-locale-sync
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          GITHUB_REPOSITORY: ${{ github.repository }}
          GITHUB_SHA: ${{ github.sha }}
          GITHUB_REF: ${{ github.ref }}
```
## Below is the response of the `Run locale sync check` steps 

```
1s
Run cd frontend
  cd frontend
  npm run check-locale-sync
  shell: /usr/bin/bash -e {0}
  env:
    GITHUB_TOKEN: ***
    GITHUB_REPOSITORY: kubestellar/ui
    GITHUB_SHA: 3883bd6d94821e021977dac4d3efe631ec490dc8
    GITHUB_REF: refs/pull/1458/merge

> kubestellarui@0.0.0 check-locale-sync
> tsx scripts/check-locale-sync.ts

🔍 Checking locale synchronization...
📋 Master locale (en) has 1621 keys
🌍 Checking 8 locale(s): de, es, fr, it, ja, pt, zh-Hans, zh-Hant
  • de: 1510 missing, 1 extra
  • es: 12 missing, 1 extra
  • fr: 11 missing, 1 extra
  • it: 1 missing, 0 extra
  • ja: 11 missing, 1 extra
  • pt: 1510 missing, 1 extra
  • zh-Hans: 140 missing, 24 extra
  • zh-Hant: 1510 missing, 1 extra
```

But commenting on PR not working ??

rishi-jat commented:
The expected behavior of the locale sync check workflow is as follows:

✅ Locale Key Synchronization

It compares all locale files in frontend/src/locales/ against the master English file (strings.en.json) to detect any missing or extra translation keys in each language.

📋 Reporting
	•	Outputs a summary in the GitHub Actions logs (e.g., how many keys are missing or extra per locale).
	•	Designed to automatically post a comment on the PR summarizing the discrepancies.

🛠️ Issue Management (Optional/Future)
	•	Can create or update a GitHub issue when translation discrepancies are found.
	•	Automatically closes the issue when all translations are fixed.

rishi-jat commented:
Currently, the sync check is running correctly and the output is visible in the logs, but the expected PR comment is not being posted. This appears to be due to a GitHub token permission issue (Resource not accessible by integration error).

rishi-jat commented:
Just to add — I’ve also updated the script to use the correct GitHub Actions output format. Previously it was using the old <<EOF syntax, which was causing issues with passing values between steps. That’s now fixed using proper key=value format with escaped newlines, so the summary and status (has_issues) should now be correctly available to later steps like the PR comment. Let me know if you’d like me to test or tweak anything further!

rishi-jat commented:
Confirmed the locale sync check works locally with env vars set.
Discrepancies are correctly detected, and has_issues: true is set.
<img width="950" height="1326" alt="image" src="https://github.com/user-attachments/assets/d54bf112-e626-4ab1-ab1f-123607e896a5" />

manzil-infinity180 commented:
What's about commenting on pr?

onkar717 commented:
@rishi-jat  any updates ?

rishi-jat commented:
@manzil-infinity180 @onkar717 All requested changes are complete PR is ready to review.

rishi-jat commented:
@manzil-infinity180 @clubanderson, could you please help me out with the PR comment part when you get a moment? Everything else is working  just stuck on this last piece.

rishi-jat commented:
@clubanderson 
<img width="1236" height="474" alt="image" src="https://github.com/user-attachments/assets/dad7eb3f-64fa-428a-8c9b-d9c7e6156069" />
The current GH_REPO_TOKEN is returning "Bad credentials" (401).

clubanderson commented:
what I see in the run of the action is in https://github.com/kubestellar/ui/actions/runs/16377668674/workflow

<img width="1125" height="538" alt="Screenshot 2025-07-22 at 11 37 35 AM" src="https://github.com/user-attachments/assets/b768f3d2-3c60-4559-93fe-ffd55d367136" />

naman9271 commented on frontend/src/locales/strings.fr.json:
```diff
@@ -2116,100 +2219,5 @@
-      "refresh": "Actualiser les Utilisateurs",
-      "addUser": "Ajouter un Utilisateur",
-      "noUsers": "Aucun Utilisateur Trouvé",
-      "noUsersDescription": "Il n'y a pas encore d'utilisateurs dans le système. Cliquez sur 'Ajouter un Utilisateur' pour créer votre premier utilisateur.",
```
why you moved this above ???
we want all translation files consistent with the `strings.en.json` one. 
and instead of adding all the missing translations in 1 PR i will suggest only fix the issue of `locale sync check` and the after that we can translate the missing strings in another PRs.

naman9271 replied:
@btwshivam what is your take on this ?

rishi-jat commented:
@clubanderson
Core functionality complete. PR comments need GH_REPO_TOKEN in main repo secrets.

naman9271 commented on .github/workflows/ci.yml:
```diff
@@ -38,6 +38,24 @@ jobs:
+          else
+            echo "❌ GH_REPO_TOKEN is empty"
+          fi
+          npm run local-sync-check
```
I can't see any changes in package.json

rishi-jat replied:
Script exists at frontend/scripts/check-locale-sync.ts and package.json line 18. Added in previous commit.
PR only adds CI step. No package.json changes needed.

rishi-jat replied:
Our PR only adds CI integration and GitHub issue creation. No package.json changes needed.

clubanderson commented:
just updated the GH_REPO_TOKEN - please test again.

clubanderson commented:
<img width="821" height="188" alt="Screenshot 2025-07-28 at 2 55 43 PM" src="https://github.com/user-attachments/assets/4acdc8f0-c2f7-497a-ad79-b88685963482" />

you already have perms for comments and pr

rishi-jat commented:
/retest

rishi-jat commented:
Hi @clubanderson,
According to the scope of issue #1444, everything is working correctly now the locale sync script runs as expected, detects discrepancies, and sets all outputs properly.

The only part still failing is the PR comment step. The CI logs show a “Permission denied” error due to token restrictions on forked PRs. Once that’s resolved, the action will be fully complete.

rishi-jat commented:
<img width="739" height="88" alt="Screenshot 2025-08-06 at 5 42 12 AM" src="https://github.com/user-attachments/assets/482cb6c7-2424-4c9f-a0f1-799da1252493" />
